aten/src/ATen/core/TensorMethods.h

```diff
@@ -69,1622 +75,4624 @@ inline void Tensor::set_data(Tensor new_data) {
 
 // all static inline to allow for inlining of the non-dynamic part of dispatch
 inline Tensor Tensor::abs() const {
+#ifdef USE_STATIC_DISPATCH
+    return TypeDefault::abs(*this);
+#else
     static auto table = globalATenDispatch().getOpTable("aten::abs(Tensor self) -> Tensor");
     return table->getOp<Tensor (const Tensor &)>(tensorTypeIdToBackend(type_id()), is_variable())(*this);
+#endif
 }
 inline Tensor & Tensor::abs_() {
+#ifdef USE_STATIC_DISPATCH
+    switch(tensorTypeIdToBackend(type_id())) {
+        case Backend::CPU:
+            return CPUType::abs_(*this);
+            break;
+        default:
+            AT_ERROR("abs_ not implemented for ", at::toString(tensorTypeIdToBackend(type_id())));
+    }
+#else
     static auto table = globalATenDispatch().getOpTable("aten::abs_(Tensor(a!) self) -> Tensor(a!)");
     return table->getOp<Tensor & (Tensor &)>(tensorTypeIdToBackend(type_id()), is_variable())(*this);
+#endif
 }
 inline Tensor Tensor::acos() const {
+#ifdef USE_STATIC_DISPATCH
+    return TypeDefault::acos(*this);
+#else
     static auto table = globalATenDispatch().getOpTable("aten::acos(Tensor self) -> Tensor");
     return table->getOp<Tensor (const Tensor &)>(tensorTypeIdToBackend(type_id()), is_variable())(*this);
+#endif
 }
 inline Tensor & Tensor::acos_() {
+#ifdef USE_STATIC_DISPATCH
+    switch(tensorTypeIdToBackend(type_id())) {
+        case Backend::CPU:
+            return CPUType::acos_(*this);
+            break;
+        default:
+            AT_ERROR("acos_ not implemented for ", at::toString(tensorTypeIdToBackend(type_id())));
+    }
+#else
     static auto table = globalATenDispatch().getOpTable("aten::acos_(Tensor(a!) self) -> Tensor(a!)");
     return table->getOp<Tensor & (Tensor &)>(tensorTypeIdToBackend(type_id()), is_variable())(*this);
+#endif
 }
 inline Tensor Tensor::add(const Tensor & other, Scalar alpha) const {
+#ifdef USE_STATIC_DISPATCH
```

Comment:
Actually, seeing this and it's similarity to the operator registration API, I got a maybe even better idea. All we need is to make sure that the `CPUType::add` kernel is present if `Tensor::add` is called. If we know it's present, we don't even need a separate static dispatcher and can just go through c10. So what if we dispatch through c10 but if it's a mobile build, we add a

    #if defined(USE_STATIC_DISPATCH)
    static auto registry = torch::RegisterOperators()...
    #endif

(or the equivalent for the intermediate dispatcher) at the top of the tensor method? Do you think that would work?

I'm not sure which approach is better but I'd like to hear your opinion.

Reply:
I feel like it's hard to factor it out because we'd be adding another layer of indirection, and in this case we'd be generalizing two cases that aren't all that similar. The static case needs the functions for each cpu backend, and the dispatcher case needs a schema/op name. Not sure if it's worth.

I think your approach of registering inside the method call will work. I think the main problem with that approach is that we need to make dispatches in `TensorMethods.h` and `Functions.h`. We'd have to deal with the additional complexity of registering in both locations. Also for the normal non-mobile case we'd still need to handle cuda registrations in separately, so we'd have registrations in 3 places. Personally I don't think this if-statement dispatch is very complex in comparison to that.

Reply:
ok let's go with the current approach for now, we'll probably have to change this later.